system/jlib/jptree.hpp

```diff
@@ -305,11 +308,15 @@ jlib_decl void mergeConfiguration(IPropertyTree & target, const IPropertyTree &
 jlib_decl IPropertyTree * loadArgsIntoConfiguration(IPropertyTree *config, const char * * argv, std::initializer_list<const char *> ignoreOptions = {});
 jlib_decl IPropertyTree * loadConfiguration(IPropertyTree * defaultConfig, const char * * argv, const char * componentTag, const char * envPrefix, const char * legacyFilename, IPropertyTree * (mapper)(IPropertyTree *), const char *altNameAttribute=nullptr);
 jlib_decl IPropertyTree * loadConfiguration(const char * defaultYaml, const char * * argv, const char * componentTag, const char * envPrefix, const char * legacyFilename, IPropertyTree * (mapper)(IPropertyTree *), const char *altNameAttribute=nullptr);
-jlib_decl IPropertyTree * queryCostsConfiguration();
+jlib_decl IPropertyTree * getCostsConfiguration();
 
 //The following can only be called after loadConfiguration has been called.  All components must call loadConfiguration().
-jlib_decl IPropertyTree & queryGlobalConfig();
-jlib_decl IPropertyTree & queryComponentConfig();
+jlib_decl IPropertyTree * getGlobalConfig();
+jlib_decl IPropertyTree * getComponentConfig();
+jlib_decl Owned<IPropertyTree> getGlobalConfigSP(); // get smart pointer
+jlib_decl Owned<IPropertyTree> getComponentConfigSP(); // get smart pointer
+jlib_decl unsigned installConfigUpdateHook(std::function<void ()> notifyFunc);
```

Comment:
I think having a notify function of the form:

void notifyFunc(IPT* oldGlobal, IPT * oldCpt, IPT * newGlobal, IPT* newCpt);

which was called before the critical section was released, would make it possible to ensure data structures are always in sync, and make it much easier to implement incremental changes.

Reply:
> I think having a notify function of the form:
> 
> void notifyFunc(IPT* oldGlobal, IPT * oldCpt, IPT * newGlobal, IPT* newCpt);
> 
> which was called before the critical section was released, would make it possible to ensure data structures are always in sync, and make it much easier to implement incremental changes.

@ghalliday - I've changed it to pass in old only, and to swap before calling, but not to release crit.
That's so that the code in the callback can see the new config via standard code (i.e. via get*Config* calls), which means that the update code will usually be the same as the original code that looks at config. settings.
It implies, intentionally, that the code is allowed to re-enter these calls (e.g. getComponentConfig()) without getting blocked, which our CS semantics allow.

Reply:
That's ok.  I had thought the cs would prevent access, but you're correct it should be allowed.